src/Parsers/ParserInsertQuery.cpp

```diff
@@ -168,6 +174,18 @@ bool ParserInsertQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
         pos = before_values;
         ParserSelectWithUnionQuery select_p;
         select_p.parse(pos, select, expected);
+        const ASTSelectWithUnionQuery ast_select = select->as<ASTSelectWithUnionQuery &>();
```

Comment:
This logic does not belong to the parser, instead it should go into the Interpreter.